installer: deps.InstallMissing should run 'apt-get update' before install on Debian/Ubuntu

## Problem

v1.98.x G2 closure integration test on a fresh Ubuntu 22.04 VM surfaced that \`deps.InstallMissing\` fails on hosts with a stale apt package index:

\`\`\`
[DEBUG] running: apt-get install -y --no-install-recommends jq socat
[WARN] DEB package install failed: apt-get install failed (exit 100): E: Unable to locate package jq
E: Unable to locate package socat
[ERROR] dependency install failed: critical dependencies still missing after install: jq, socat
[ERROR] phase Prepare failed: FAILED_RENDER
\`\`\`

## Root cause

\`internal/installer/deps/deps.go\` calls \`apt-get install\` directly without a preceding \`apt-get update\`. On fresh VMs (cloud-init, minimal images, old snapshots), the apt index can be stale/empty — the install command can't find any package.

## Why it didn't show up before

Ubuntu 24.04 / Debian 12 / Debian 13 test hosts in the same run all had recent apt-updates from their base images or prior runs. The specific Ubuntu 22.04 provisioning template delivered a VM with an older cached apt index that had never been refreshed.

## Proposed fix

In \`internal/installer/deps/deps.go\` (Debian/Ubuntu branch), run \`apt-get update\` before \`apt-get install\`:

\`\`\`go
if isDebianFamily(distro) {
    // Refresh package index. apt-get install on a stale index fails
    // with "Unable to locate package" on fresh VMs.
    exec.RunTimeout(60*time.Second, "apt-get", "update")
    res := exec.Run("apt-get", "install", "-y", "--no-install-recommends", pkgs...)
    // ...
}
\`\`\`

\`apt-get update\` failure should be non-fatal (log and proceed) — if the index is already current, update will be a no-op anyway. The authoritative failure is \`apt-get install\` returning non-zero.

## Impact

- **Source install on freshly-provisioned apt-based VMs**: currently fails FAILED_RENDER
- **Package installs (RPM/DEB)**: not affected — deps are resolved by the package manager itself via dependencies in the \`.deb\`/\`.rpm\` metadata
- **Source install on apt-based systems with recent apt-update**: not affected (current behavior)

## Workaround

Operator runs \`apt-get update\` manually before \`./install.sh\`. Documentable but not a good UX.

## Classification

Non-blocking for v1.98.x closure — only affects source-install path on a specific host condition that's avoidable by a 1-line operator command. Worth fixing in v1.99 or a v1.98.x point release.

## Related

- PR #462 — Go-side source-install support (where this code path lives)
- v1.98.x G2 closure integration test (Ubuntu 22.04 run at 2026-04-19T13:36:22Z)

--- a/internal/installer/deps/deps.go
+++ b/internal/installer/deps/deps.go
@@ -143,10 +143,34 @@ func installRPM(exec executor.Executor, pkgs []string, log *logging.Logger) erro
 	return nil
 }
 
+// aptGetUpdate refreshes the apt package index. Called as a preflight
+// before apt-get install so a stale index does not cause a misleading
+// "Unable to locate package" failure on fresh VMs (issue #467).
+//
+// Operator policy: deterministic correctness — if the index refresh
+// itself fails, surface a clear preflight error rather than continuing
+// into the install command (which would fail less readably).
+func aptGetUpdate(exec executor.Executor, log *logging.Logger) error {
+	log.Debug("running: apt-get update (preflight)")
+	res := exec.RunTimeout(60*time.Second, "apt-get", "update")
+	if res.ExitCode != 0 {
+		return fmt.Errorf("apt-get update preflight failed (exit %d): %s", res.ExitCode, res.Stderr)
+	}
+	return nil
+}
+
 // installDEB uses apt-get to install packages.
 func installDEB(exec executor.Executor, pkgs []string, log *logging.Logger) error {
 	if !exec.CommandExists("apt-get") {
 		return fmt.Errorf("apt-get not available")
+	}
+
+	// Preflight: refresh apt index before install. Required on fresh VMs
+	// with stale apt cache; otherwise apt-get install fails with
+	// "Unable to locate package". Fail-fast on update error so the
+	// operator sees the real cause (issue #467).
+	if err := aptGetUpdate(exec, log); err != nil {
+		return err
 	}
 
 	args := append([]string{"install", "-y", "--no-install-recommends"}, pkgs...)

--- a/internal/installer/deps/deps_test.go
+++ b/internal/installer/deps/deps_test.go
@@ -216,3 +216,130 @@ func TestIsRPMFamily(t *testing.T) {
 		}
 	}
 }
+
+// PR-P4 (issue #467): apt-get update preflight before apt-get install.
+
+// TestInstallMissing_DEB_RunsAptGetUpdateBeforeInstall verifies that the
+// DEB path issues `apt-get update` BEFORE `apt-get install` so a stale
+// apt index does not cause a misleading "Unable to locate package".
+func TestInstallMissing_DEB_RunsAptGetUpdateBeforeInstall(t *testing.T) {
+	mock := executor.NewMockExecutor()
+	for _, d := range requiredDeps {
+		if d.cmd != "jq" {
+			mock.ExistingCommands[d.cmd] = true
+		}
+	}
+	mock.ExistingCommands["apt-get"] = true
+
+	distro := &detect.DistroInfo{ID: "ubuntu", VersionID: "22.04"}
+	// jq remains "missing" so the post-install check returns critical-still-missing,
+	// but that's fine — we only assert the call ordering.
+	_ = InstallMissing(mock, distro, newTestLogger())
+
+	updateIdx := -1
+	installIdx := -1
+	for i, cmd := range mock.Commands {
+		if cmd.Name != "apt-get" {
+			continue
+		}
+		if len(cmd.Args) > 0 && cmd.Args[0] == "update" && updateIdx == -1 {
+			updateIdx = i
+		}
+		if len(cmd.Args) > 0 && cmd.Args[0] == "install" && installIdx == -1 {
+			installIdx = i
+		}
+	}
+	if updateIdx == -1 {
+		t.Fatal("expected apt-get update to be called as preflight")
+	}
+	if installIdx == -1 {
+		t.Fatal("expected apt-get install to be called")
+	}
+	if updateIdx >= installIdx {
+		t.Errorf("apt-get update must precede apt-get install (updateIdx=%d, installIdx=%d)", updateIdx, installIdx)
+	}
+}
+
+// TestInstallMissing_DEB_AptGetUpdateFailureIsFatal verifies that when the
+// apt-get update preflight fails, InstallMissing returns a clear preflight
+// error AND apt-get install is NOT attempted.
+func TestInstallMissing_DEB_AptGetUpdateFailureIsFatal(t *testing.T) {
+	mock := executor.NewMockExecutor()
+	for _, d := range requiredDeps {
+		if d.cmd != "jq" {
+			mock.ExistingCommands[d.cmd] = true
+		}
+	}
+	mock.ExistingCommands["apt-get"] = true
+
+	// Seed apt-get update to fail (simulates network-down / repo unreachable).
+	mock.RunResults["apt-get:update"] = executor.Result{
+		ExitCode: 100,
+		Stderr:   "E: Could not resolve 'archive.ubuntu.com'",
+	}
+
+	distro := &detect.DistroInfo{ID: "debian", VersionID: "12"}
+	err := InstallMissing(mock, distro, newTestLogger())
+	if err == nil {
+		t.Fatal("expected error when apt-get update preflight fails (critical dep missing path)")
+	}
+
+	// Find apt-get update + assert apt-get install was NOT attempted.
+	sawUpdate := false
+	sawInstall := false
+	for _, cmd := range mock.Commands {
+		if cmd.Name != "apt-get" {
+			continue
+		}
+		if len(cmd.Args) > 0 && cmd.Args[0] == "update" {
+			sawUpdate = true
+		}
+		if len(cmd.Args) > 0 && cmd.Args[0] == "install" {
+			sawInstall = true
+		}
+	}
+	if !sawUpdate {
+		t.Fatal("expected apt-get update to be attempted")
+	}
+	if sawInstall {
+		t.Error("apt-get install must NOT run after apt-get update preflight failure")
+	}
+
+	// Direct helper check: aptGetUpdate must surface a clear preflight error.
+	directErr := aptGetUpdate(mock, newTestLogger())
+	if directErr == nil {
+		t.Fatal("expected aptGetUpdate to return an error when apt-get exits non-zero")
+	}
+	msg := directErr.Error()
+	if !strings.Contains(msg, "apt-get update preflight failed") {
+		t.Errorf("expected error to contain 'apt-get update preflight failed', got: %v", directErr)
+	}
+	if !strings.Contains(msg, "100") {
+		t.Errorf("expected error to contain exit code 100, got: %v", directErr)
+	}
+	if !strings.Contains(msg, "archive.ubuntu.com") {
+		t.Errorf("expected error to contain stderr text, got: %v", directErr)
+	}
+}
+
+// TestInstallMissing_RPM_DoesNotRunAptGetUpdate verifies that the RPM/DNF
+// path is unaffected by the PR-P4 preflight (apt-get update only applies
+// to the DEB family).
+func TestInstallMissing_RPM_DoesNotRunAptGetUpdate(t *testing.T) {
+	mock := executor.NewMockExecutor()
+	for _, d := range requiredDeps {
+		if d.cmd != "socat" {
+			mock.ExistingCommands[d.cmd] = true
+		}
+	}
+	mock.ExistingCommands["dnf"] = true
+
+	distro := &detect.DistroInfo{ID: "almalinux", VersionID: "9"}
+	_ = InstallMissing(mock, distro, newTestLogger())
+
+	for _, cmd := range mock.Commands {
+		if cmd.Name == "apt-get" {
+			t.Errorf("RPM path must not call apt-get (got: %s %v)", cmd.Name, cmd.Args)
+		}
+	}
+}

